program/src/state.rs

```diff
@@ -0,0 +1,622 @@
+    Activated,
+    Deactivating { remaining_blocks: usize },
+    Deactivated,
+}
+
+/// Address lookup table metadata
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone, AbiExample)]
+pub struct LookupTableMeta {
+    // [Core BPF]: TODO: `Clock` instead of `SlotHashes`.
+    /// Lookup tables cannot be closed until the deactivation slot is
+    /// no longer "recent" (not accessible in the `SlotHashes` sysvar).
+    pub deactivation_slot: Slot,
+    /// The slot that the table was last extended. Address tables may
+    /// only be used to lookup addresses that were extended before
+    /// the current bank's slot.
+    pub last_extended_slot: Slot,
+    /// The start index where the table was last extended from during
+    /// the `last_extended_slot`.
+    pub last_extended_slot_start_index: u8,
+    /// Authority address which must sign for each modification.
+    pub authority: Option<Pubkey>,
+    // Padding to keep addresses 8-byte aligned
+    pub _padding: u16,
+    // Raw list of addresses follows this serialized structure in
+    // the account's data, starting from `LOOKUP_TABLE_META_SIZE`.
+}
+
+impl Default for LookupTableMeta {
+    fn default() -> Self {
+        Self {
+            deactivation_slot: Slot::MAX,
+            last_extended_slot: 0,
+            last_extended_slot_start_index: 0,
+            authority: None,
+            _padding: 0,
+        }
+    }
+}
+
+impl LookupTableMeta {
+    pub fn new(authority: Pubkey) -> Self {
+        LookupTableMeta {
+            authority: Some(authority),
+            ..LookupTableMeta::default()
+        }
+    }
+
+    /// Returns whether the table is considered active for address lookups
+    pub fn is_active(&self, current_slot: Slot) -> bool {
+        match self.status(current_slot) {
+            LookupTableStatus::Activated => true,
+            LookupTableStatus::Deactivating { .. } => true,
+            LookupTableStatus::Deactivated => false,
+        }
+    }
+
+    // [Core BPF]: This function has been modified from its legacy built-in
+    // counterpart to no longer use the `SlotHashes` sysvar, since it is not
+    // available for BPF programs. Instead, it uses the `current_slot`
+    // parameter to calculate the table's status.
+    // This will no longer consider the case where a slot has been skipped
+    // and no block was produced.
+    // If it's imperative to ensure we are only considering slots where blocks
+    // were created, then we'll need to revisit this function, and possibly
+    // provide the `SlotHashes` account so we can reliably check slot hashes.
+    /// Return the current status of the lookup table
+    pub fn status(&self, current_slot: Slot) -> LookupTableStatus {
+        if self.deactivation_slot == Slot::MAX {
+            LookupTableStatus::Activated
+        } else if self.deactivation_slot == current_slot {
+            LookupTableStatus::Deactivating {
+                remaining_blocks: MAX_ENTRIES.saturating_add(1),
```

Comment:
This should be `MAX_ENTRIES` I believe, based on the other logic which does `position >= MAX_ENTRIES`. That means that 512 slots later, `calculate_slot_position` will return `None` and allow closing.

Reply:
I don't think this check should change from the original implementation. It's still considering the current slot in the exact same way. In the `SlotHashes` case you'd have `512` slots in the sysvar plus the current slot that hasn't been added yet.

https://github.com/solana-labs/solana/blob/e81ecfbe9fbc2b6f95e6f8f5f88cb1af39cb24aa/sdk/program/src/address_lookup_table/state.rs#L82-L85

I'm thinking maybe instead that check should be `position > MAX_ENTRIES`, since that's how the `SlotHashes` lookup would work for a prior deactivation slot.

Reply:
* if current_slot = deactivation_slot, then it'll say wait 513 slots (512 + 1)
* if current_slot = deactivation_slot + 1, then `calculate_slot_position` will return 1, and say you need to wait 511 slots

Something's off, right?

Reply:
And to keep going, if current_slot = deactivation_slot + 511, then `calculate_slot_position` will return `511`, and say wait 1 more slot

If current_slot = deactivation_slot + 512, then `calculate_slot_position` will say `None`.

Did I get anything wrong?

Reply:
Ahh, I think this is what I'm missing:

```rust
let position = current_slot.saturating_sub(1).saturating_sub(*target_slot);
```

I added a commit with some comments explaining the reasoning for subtracting `1` there. TL/DR we have to account for the current slot not being present in the `SlotHashes`.

Reply:
To be honest, I think it'll be simpler to remove all of the add / subtract of 1. What do you think?

Reply:
So just make cooldown one slot shorter then? I guess it doesn't really matter since the behavior is changing already, and cooldown is now going to be shorter anyway, by way of slots vs blocks. I think it's OK.

Reply:
Yeah exactly, just to avoid any future off-by-one errors

Reply:
Alrighty, will do!